### PR TITLE
 Testing routes and constraints without making actual requests.

### PR DESCRIPTION
**Pull request by varshavaradarajan**
https://github.com/gocd/gocd/issues/2532

### CODE REVIEW

**Comment by ketan** [comment]
The approach looks good to me, can we move these into specific controllers instead?

**Comment by varshavaradarajan** [comment]
Well, I know the specs aren't exactly routing but is there any reason not to have this structure?

```
└── routing
    ├── api
    ├── api_v1
    └── api_v2
```

File: server/webapp/WEB-INF/rails.new/spec/controllers/api_v1/admin/environments_controller_spec.rb [hunk 1]
```diff
@@ -101,6 +121,42 @@
...
+
+    describe :route do
+      describe :with_header do
+        before :each do
+          Rack::MockRequest::DEFAULT_ENV["HTTP_ACCEPT"] = "application/vnd.go.cd.v1+json"
+        end
+        after :each do
+          Rack::MockRequest::DEFAULT_ENV = {}
+        end
+        it 'should route to show action of environments controller for alphanumeric environment name' do
```
Comment (ketan) [inline server/webapp/WEB-INF/rails.new/spec/controllers/api_v1/admin/environments_controller_spec.rb]: You can reduce a lot of dup if you change most of this code to —

```ruby
%w(foo123 foo.123 foo-123).each do |name|
  it "should route ... to ... with param name #{name}" do
    expect(get: ".../#{name}").to route_to(action: ..., controller: ..., name: ...)
  end
end

```

Comment (maheshp) [inline server/webapp/WEB-INF/rails.new/spec/controllers/api_v1/admin/environments_controller_spec.rb]: The only drawback I see with this approach is that the testing scenario  is not explicit,

eg: ```it('should route to ... for pipeline name with dot')``` somebody will have to see the test data to understand the scenarios being tested.

File: server/webapp/WEB-INF/rails.new/config/routes.rb [hunk 2]
```diff
@@ -26,7 +26,7 @@
     PIPELINE_LOCATOR_CONSTRAINTS = {:pipeline_name => PIPELINE_NAME_FORMAT, :pipeline_counter => PIPELINE_COUNTER_FORMAT}
     STAGE_LOCATOR_CONSTRAINTS = {:stage_name => STAGE_NAME_FORMAT, :stage_counter => STAGE_COUNTER_FORMAT}.merge(PIPELINE_LOCATOR_CONSTRAINTS)
     ENVIRONMENT_NAME_CONSTRAINT = {:name => ENVIRONMENT_NAME_FORMAT}
-    PLUGIN_ID_FORMAT = /[a-zA-Z0-9\-_.]+/
+    PLUGIN_ID_FORMAT = /[\w\-.]+/
```
Comment (ketan) [inline server/webapp/WEB-INF/rails.new/config/routes.rb]: See https://github.com/gocd/gocd/blob/242d113bb7724911ee0c7166b200d56a86f3ada3/plugin-infra/go-plugin-api/resources/plugin-descriptor.xsd#L33-L33

File: server/webapp/WEB-INF/rails.new/spec/controllers/api_v1/admin/environments_controller_spec.rb [hunk 3]
```diff
@@ -54,6 +54,26 @@
...
     end
+
+    describe :route do
+      describe :with_header do
+        before :each do
+          Rack::MockRequest::DEFAULT_ENV["HTTP_ACCEPT"] = "application/vnd.go.cd.v1+json"
+        end
+        after :each do
+          Rack::MockRequest::DEFAULT_ENV = {}
+        end
```
Comment (ketan) [inline server/webapp/WEB-INF/rails.new/spec/controllers/api_v1/admin/environments_controller_spec.rb]: `DEFAULT_ENV` is supposed to contain other keys as well — https://github.com/gocd/gocd/blob/f53cc516f5ef04295fd03dfa9dd8160b0f0745cd/server/webapp/WEB-INF/rails.new/vendor/bundle/jruby/1.9/gems/rack-1.5.5/lib/rack/mock.rb#L43-L50

Might as well include a helper module —

```ruby
module ApiHeaderSetupTeardown
  def self.included(klass)
    klass.class_eval do
      before :each do
        @default_env = Rack::MockRequest::DEFAULT_ENV.dup
        Rack::MockRequest::DEFAULT_ENV.merge!({'HTTP_ACCEPT' => "application/vnd.go.cd.v#{supported_version}+json"})
      end
      after :each do
        Rack::MockRequest::DEFAULT_ENV = @default_env
      end
    end
  end
end

describe FooController do
  include ApiHeaderSetupTeardown
  def supported_version
    1
  end

  it 'should foo' do
    ...
  end
end
```

**Pull request merged**